### DISCUSSION TITLE
Duplication between prisma schema and models

### DISCUSSION THREAD

**Discussion by Skona27** [fivethree-team/nestjs-prisma-client-starter#446] [Q&A] [answered]
Hi!
First of all thanks for this starter. It is really awesome 😃 

I've noticed that some of the code is duplicated between prisma schema definitions and nest js models.
In `prisma.schema` we have a `User` definition, which looks like below:
```
model User {
  id        String   @id @default(cuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  email     String   @unique
  password  String
  firstname String?
  lastname  String?
  posts     Post[]
  role      Role
}
```
And also in `user.model` we define the same 'shape' but with typescript:
```
@ObjectType()
export class User extends BaseModel {
  email: string;
  firstname?: string;
  lastname?: string;
  role: Role;
  @HideField()
  password: string;
}
```
Why is that necessary? Can't we somehow generate schema from models, or other way around?

**Comment by marcjulian** [comment, marked as answer]
@Skona27 Hi, thanks 👍 I hope I can answer your question.

Yes currently the code requires duplicates between the prisma schema and the graphql schema.

1. Database schema: `schema.prisma` describes the model and thus the table created for this model.
2. GraphQL schema: `user.model.ts` and any other models describes the response value of your `user` type for your GraphQL queries

How to resolve the duplicates for now?
It is currently **necessary** to have duplicate code once in `schema.prisma` and once for your GraphQL schema.
  
One idea would be to implement `User` from prisma client for your graphql model to always keep those prisma and graphql models in sync.

```ts
import { User } from '@prisma/client'; 
import { PostModel } from '../post/post.model.ts';

@ObjectType('User')
export class UserModel extends BaseModel implements User {
  email: string;
  @Field({ nullable: true })
  firstname: string;
  @Field({ nullable: true })
  lastname: string;
  role: Role;
  @HideField()
  password: string;
  posts: PostModel[];
}
```

Another option might be in the feature [typegraphql-prisma](https://github.com/MichalLytek/typegraphql-prisma/issues/1)

> And in the near feature - when Prisma SDK will be ready - the typegraphql-prisma integration will also allow to use a code-first approach to build a schema.prisma and GraphQL schema at once, using classes with decorators as a single source of truth. Stay tuned! 💪

There is also an typegraphql integration for nestjs [typegraphql-nestjs](https://github.com/MichalLytek/typegraphql-nestjs) with some [caveats](https://github.com/MichalLytek/typegraphql-nestjs#caveats).

**Reply by marcjulian** [reply]
Anothre option could be https://github.com/unlight/prisma-nestjs-graphql which is similar to https://github.com/MichalLytek/typegraphql-prisma, but works with NestJS GraphQL Module.

**Comment by Skona27** [comment]
@marcjulian 
Thanks for that answer! Everything is clear now

**Comment by orenmizr** [comment]
its been a while since asked. is there now a way to not have duplicate data ? 

i thought it be easier that classes generate prisma DSL but it does not seem to be a thing.

**Reply by lucassith** [reply]
Yes, there a way to avoid duplicated data, just don't use prisma with NestJS. I know this is probably the most disappointing answer, but it seems like the technology isn't really compatible with each other and it's just getting exclusive set.

TypeORM seems to be more fitting package (but less hyped atm).

**Comment by hmiiro** [comment]
With prisma, I use the generated models(types) from prisma client. Why need
to generate models again?
